File: server/src/cleanup.ts
import { eq, and, lt, isNotNull, sql, desc } from 'drizzle-orm'
import { getDb, schema } from './db/index.js'
import { deleteFile } from './storage.js'

const DEFAULT_INTERVAL_MS = 60 * 60 * 1000 // 1 hour

let timer: ReturnType<typeof setInterval> | null = null
let running = false

/** Get configured cleanup interval from system_settings */
async function getConfiguredIntervalMs(): Promise<number> {
  const db = getDb()
  const row = await db.select().from(schema.systemSettings)
    .where(eq(schema.systemSettings.key, 'cleanup_interval_minutes'))
    .limit(1)
  if (!row[0]) return DEFAULT_INTERVAL_MS
  const minutes = parseInt(row[0].value)
  if (isNaN(minutes) || minutes <= 0) return DEFAULT_INTERVAL_MS
  return minutes * 60 * 1000
}

/** Delete expired attachments from DB and disk */
async function cleanupExpiredAttachments(): Promise<number> {
  const db = getDb()
  const now = new Date()

  // Get filenames for disk cleanup, then bulk delete from DB
  const expired = await db.select({
    id: schema.attachments.id,
    filename: schema.attachments.filename,
  }).from(schema.attachments)
    .where(and(
      isNotNull(schema.attachments.expiresAt),
      lt(schema.attachments.expiresAt, now)
    ))

  if (expired.length === 0) return 0

  // Delete files from disk
  for (const att of expired) {
    await deleteFile(att.filename)
  }

  // Bulk delete from DB
  const ids = expired.map(a => a.id)
  await db.delete(schema.attachments)
    .where(sql`${schema.attachments.id} IN (${sql.join(ids.map(id => sql`${id}`), sql`, `)})`)

  return expired.length
}

/** Delete expired messages for all users who have messageExpiration set */
async function cleanupExpiredMessages(): Promise<number> {
  const db = getDb()

  const usersWithExpiration = await db.select().from(schema.userSettings)
    .where(sql`${schema.userSettings.messageExpiration} > 0`)

  if (usersWithExpiration.length === 0) return 0

  let totalDeleted = 0
  for (const settings of usersWithExpiration) {
    const cutoff = new Date()
    cutoff.setDate(cutoff.getDate() - settings.messageExpiration)

    // Bulk delete instead of SELECT→loop→DELETE
    const result = await db.delete(schema.messages)
      .where(and(
        eq(schema.messages.userId, settings.userId),
        lt(schema.messages.createdAt, cutoff)
      ))

    totalDeleted += (result as any).rowsAffected ?? 0
  }

  return totalDeleted
}

/** Enforce per-user message limits */
async function enforceUserMessageLimits(): Promise<number> {
  const db = getDb()

  const maxRow = await db.select().from(schema.systemSettings)
    .where(eq(schema.systemSettings.key, 'max_messages_per_user'))
    .limit(1)
  const maxMessages = maxRow[0] ? parseInt(maxRow[0].value) : 1000

  const userIds = await db.select({ userId: schema.messages.userId })
    .from(schema.messages)
    .where(sql`${schema.messages.userId} IS NOT NULL`)
    .groupBy(schema.messages.userId)

  let totalTrimmed = 0
  for (const { userId } of userIds) {
    if (!userId) continue

    const countRow = await db.select({ count: sql<number>`count(*)` })
      .from(schema.messages)
      .where(eq(schema.messages.userId, userId))
    const count = countRow[0]?.count ?? 0

    if (count <= maxMessages) continue

    const excess = count - maxMessages
    // Bulk delete using subquery instead of loop
    await db.run(sql`
      DELETE FROM messages
      WHERE id IN (
        SELECT id FROM messages
        WHERE user_id = ${userId}
        ORDER BY created_at ASC
        LIMIT ${excess}
      )
    `)
    totalTrimmed += excess
  }

  return totalTrimmed
}

/** Run all cleanup tasks and record results */
async function runCleanup(): Promise<void> {
  if (running) return
  running = true

  const db = getDb()
  const startedAt = new Date()

  // Insert a running log entry
  const [logEntry] = await db.insert(schema.cleanupLogs).values({
    startedAt,
    status: 'running',
  }).returning()

  let expiredAttachments = 0
  let expiredMessages = 0
  let trimmedMessages = 0
  let error: string | null = null

  try {
    expiredAttachments = await cleanupExpiredAttachments()
    expiredMessages = await cleanupExpiredMessages()
    trimmedMessages = await enforceUserMessageLimits()
  } catch (err) {
    error = err instanceof Error ? err.message : String(err)
    console.error('[cleanup] Error during cleanup:', err)
  }

  const finishedAt = new Date()
  const durationMs = finishedAt.getTime() - startedAt.getTime()

  await db.update(schema.cleanupLogs)
    .set({
      finishedAt,
      durationMs,
      status: error ? 'error' : 'success',
      expiredAttachments,
      expiredMessages,
      trimmedMessages,
      error,
    })
    .where(eq(schema.cleanupLogs.id, logEntry.id))

  const parts: string[] = []
  if (expiredAttachments > 0) parts.push(`${expiredAttachments} attachments`)
  if (expiredMessages > 0) parts.push(`${expiredMessages} expired messages`)
  if (trimmedMessages > 0) parts.push(`${trimmedMessages} trimmed messages`)

  if (parts.length > 0) {
    console.log(`[cleanup] Done in ${durationMs}ms: ${parts.join(', ')}`)
  } else {
    console.log(`[cleanup] Done in ${durationMs}ms: nothing to clean`)
  }

  running = false
}

/** Start the cleanup scheduler */
export async function startCleanupScheduler() {
  // Run immediately on startup
  runCleanup().catch(err => {
    console.error('[cleanup] Initial run failed:', err)
  })

  // Schedule periodic runs
  const scheduleNext = async () => {
    const intervalMs = await getConfiguredIntervalMs()
    if (timer) clearInterval(timer)
    timer = setInterval(() => {
      runCleanup().catch(err => {
        console.error('[cleanup] Periodic run failed:', err)
      })
    }, intervalMs)
    console.log(`[cleanup] Scheduler started (interval: ${Math.round(intervalMs / 60000)}min)`)
  }

  await scheduleNext()

  // Re-read interval every 5 minutes in case admin changed it
  setInterval(() => {
    scheduleNext().catch(() => {})
  }, 5 * 60 * 1000)
}

/** Get cleanup logs (paginated) */
export async function getCleanupLogs(page = 1, pageSize = 20) {
  const db = getDb()
  const offset = (page - 1) * pageSize

  const items = await db.select().from(schema.cleanupLogs)
    .orderBy(desc(schema.cleanupLogs.startedAt))
    .limit(pageSize)
    .offset(offset)

  const countRow = await db.select({ count: sql<number>`count(*)` })
    .from(schema.cleanupLogs)
  const total = countRow[0]?.count ?? 0

  return { items, total, page, pageSize }
}

File: rust-server/server/src/worker/cleanup.rs
use sqlx::SqlitePool;
use std::sync::Arc;
use tokio::time::{sleep, Duration};

use crate::config::Config;

/// Start the cleanup scheduler (runs every hour)
pub async fn start(pool: SqlitePool, _config: Arc<Config>) {
    tracing::info!("[cleanup] Starting cleanup scheduler");

    // Run immediately on startup
    if let Err(e) = run_cleanup(&pool).await {
        tracing::error!("[cleanup] Initial cleanup error: {e}");
    }

    loop {
        // Run cleanup every hour
        sleep(Duration::from_secs(3600)).await;

        if let Err(e) = run_cleanup(&pool).await {
            tracing::error!("[cleanup] Error: {e}");
        }
    }
}

async fn run_cleanup(pool: &SqlitePool) -> anyhow::Result<()> {
    let now = chrono::Utc::now().timestamp();
    let started_at = now;

    tracing::info!("[cleanup] Starting cleanup run");

    // 1. Clean up expired attachments
    let expired_files: Vec<(String,)> = sqlx::query_as(
        "SELECT filename FROM attachments WHERE expires_at IS NOT NULL AND expires_at <= ?"
    )
    .bind(now)
    .fetch_all(pool)
    .await?;

    let expired_attachments = expired_files.len() as i64;

    for (filename,) in &expired_files {
        let path = std::path::Path::new("data/uploads").join(filename);
        if path.exists() {
            if let Err(e) = tokio::fs::remove_file(&path).await {
                tracing::warn!("[cleanup] Failed to delete file {filename}: {e}");
            }
        }
    }

    if expired_attachments > 0 {
        sqlx::query("DELETE FROM attachments WHERE expires_at IS NOT NULL AND expires_at <= ?")
            .bind(now)
            .execute(pool)
            .await?;
    }

    // 2. Clean up expired messages (per user settings)
    let expired_messages = cleanup_expired_messages(pool, now).await?;

    // 3. Trim old messages per user limits
    let trimmed_messages = trim_user_messages(pool).await?;

    // 4. Clean up old delivered push messages (older than 7 days)
    let old_push = now - 604800;
    sqlx::query("DELETE FROM push_messages WHERE delivered = 1 AND created_at < ?")
        .bind(old_push)
        .execute(pool)
        .await?;

    // 5. Clean up old app logs (older than 30 days)
    let old_logs = now - 2592000;
    sqlx::query("DELETE FROM app_logs WHERE created_at < ?")
        .bind(old_logs)
        .execute(pool)
        .await?;

    let finished_at = chrono::Utc::now().timestamp();
    let duration_ms = (finished_at - started_at) * 1000;

    // Log cleanup run
    sqlx::query(
        r#"INSERT INTO cleanup_logs (started_at, finished_at, duration_ms, status, expired_attachments, expired_messages, trimmed_messages)
           VALUES (?, ?, ?, 'success', ?, ?, ?)"#
    )
    .bind(started_at)
    .bind(finished_at)
    .bind(duration_ms)
    .bind(expired_attachments)
    .bind(expired_messages)
    .bind(trimmed_messages)
    .execute(pool)
    .await?;

    tracing::info!(
        "[cleanup] Done: {expired_attachments} expired attachments, {expired_messages} expired messages, {trimmed_messages} trimmed messages ({duration_ms}ms)"
    );

    Ok(())
}

async fn cleanup_expired_messages(pool: &SqlitePool, now: i64) -> anyhow::Result<i64> {
    // Get users with message expiration settings
    let users: Vec<(i64, i64)> = sqlx::query_as(
        "SELECT user_id, message_expiration FROM user_settings WHERE message_expiration > 0"
    )
    .fetch_all(pool)
    .await?;

    let mut total = 0i64;

    for (user_id, expiry_days) in users {
        let cutoff = now - (expiry_days * 86400);
        let result = sqlx::query("DELETE FROM messages WHERE user_id = ? AND created_at < ?")
            .bind(user_id)
            .bind(cutoff)
            .execute(pool)
            .await?;
        total += result.rows_affected() as i64;
    }

    Ok(total)
}

async fn trim_user_messages(pool: &SqlitePool) -> anyhow::Result<i64> {
    let max_messages: i64 = sqlx::query_scalar(
        "SELECT CAST(value AS INTEGER) FROM system_settings WHERE key = 'max_messages_per_user'"
    )
    .fetch_optional(pool)
    .await?
    .unwrap_or(1000);

    let users: Vec<(i64,)> = sqlx::query_as("SELECT DISTINCT user_id FROM messages WHERE user_id IS NOT NULL")
        .fetch_all(pool)
        .await?;

    let mut total = 0i64;

    for (user_id,) in users {
        let count: (i64,) = sqlx::query_as("SELECT COUNT(*) FROM messages WHERE user_id = ?")
            .bind(user_id)
            .fetch_one(pool)
            .await?;

        if count.0 > max_messages {
            let excess = count.0 - max_messages;
            sqlx::query(
                "DELETE FROM messages WHERE id IN (SELECT id FROM messages WHERE user_id = ? ORDER BY created_at ASC LIMIT ?)"
            )
            .bind(user_id)
            .bind(excess)
            .execute(pool)
            .await?;
            total += excess;
        }
    }

    Ok(total)
}
